### ISSUE TITLE
Low pass filter definition

### ISSUE THREAD

**Issue by nshervt** [alelab-upenn/graph-scattering-transforms#1]
Hello, I was wondering why the low-pass filter `U` in the diffusion wavelet is defined as a vector of shape `(N)`, N being the number of nodes:

`self.U = d/np.linalg.norm(d, 1)
`

I was expecting `U` to be a square matrix. I found that [Zou and Lerman](https://arxiv.org/abs/1804.00099) (which you have cited) are also defining `U` as a square matrix of size `(N, N)`.
`U = np.linalg.matrix_power(T, t)
`

The fact that you are using a vector also gives a different number of output features. 

Same question goes for tight Hann wavelets where low-pass filter `U` is

`self.U = (1/self.N) * np.ones(self.N)`

In scattering transform, scaling function is used as low pass filter during scattering transform to capture low frequency or average behavior of the signal. I was expecting that you use equation (13) of [Shuman et. al.](https://ieeexplore.ieee.org/abstract/document/7088640) as low-pass filter, and use equation (12) as wavelets, but it seems that you have used both relations as wavelets and used `self.U = (1/self.N) * np.ones(self.N)` as low-pass filter. 

1. Isn't scaling supposed to be the low pass filter that computes invariants and wavelets the high pass filters? 

2. using a vector as opposed to a squared matrix makes your implementation of Diffusive wavelets have different dimensionality of output features in comparison with [Zou and Lerman](https://arxiv.org/abs/1804.00099). Which one should I trust?

**Comment by fgfgama** [comment]
Hello, nshervt!

1. You're correct: the operator U is used to compute an invariant out of the graph convolution operation (which is, in itself, equivariant). We use a low-pass summary (average) to do this (<code>self.U = (1/self.N) * np.ones(self.N)</code> for the Hann and Cubic wavelets, <code>self.U = d/np.linalg.norm(d,1)</code> for the Diffusion wavelets). Since these are average operators they are low-pass filters as well. The wavelets, as you correctly point out, capture high-pass information (inasmuch as their frequency response at zero-frequency is zero).

2. The use of a vector to compute an invariant is in analogy to the operator S[p] of the work by Bruna and Mallat (2013), see page 1874, right before Section 2.3. This operator computes an l-1 norm (or, equivalently, an absolute value integrated over the domain), producing a single scalar: the scattering coefficient obtained at each output -after applying the operator S[p] in Bruna or Mallat, or the operator U in our case- is a scalar. Then, the GST is the set of these scalars, collected at each layer, for each scale (see below eq. (6), page 1785). This has the main advantage of rendering the number of GST coefficients independent of the number of nodes (it only depends on the number of layers L and the number of scales J), so that the output representation is independent of the size of the input data, as is intended in Bruna and Mallat (2013). Additionally, we would like to note that it is not possible to design a low-pass **graph** filter that is invariant -they are, at most, equivariant-.

**Comment by nshervt** [comment]
Thank you for the detailed explanation @fgfgama! 

One point still baffles me: If you are using `self.U = (1/self.N) * np.ones(self.N)` for low-pass filtering, why are you still defining scaling function `psi = R * 0.25 + R/2 * 0.25 - sumPsiSquared `? And more importantly, why are you using it as a high-pass filter? Correct me if I'm wrong, but you store scaling function in variable `H`, which is later used as high-pass filter (even naming it as `psi`, rather than `phi`, suggests that you are treating it as wavelet rather than scaling function). Is there a reason for that?

**Comment by fgfgama** [comment]
That's a very good observation, @nshervt ! Thanks for bringing it up.

In the building of the GST, we have three components: a filter bank h_j, a pointwise nonlinearity rho, and a summary operator U (the key aspect here, is that it is a summary operator, so that the number of coefficients in the GST representation -the phi- don't blow up with the number of nodes). And this is all we need to define a GST.

Now, we want two more things: to be able to say something about the stability of GSTs, and to be able to show that the GST representation is useful. In order to achieve this, we need to start imposing conditions on these three elements.

For stability, we require that all the filters h_j in the bank satisfy the integral Lipschitz condition. We know wavelets do this. We also need pointwise nonlinearities that are Lipschitz continuous, ideally with Lipschitz constant equal to 1, and the absolute value does this.

Now that we have fixed that we will be working with wavelets, so that the integral Lipschitz conditions holds, and with absolute values, we want the GST to extract useful information. We observe that wavelet filter banks extract information in different frequency bands, with different bandwidths, and that the absolute value spreads the information throughout the frequency spectrum, so that the output of each wavelets give a different aspect (a different feature) of the signal, restricted to the corresponding frequency band. Since most (if not all) of the wavelets are bandpass filters, we don't want to leave the DC component of the signal go to waste, so we force the summary operator be a low-pass as well.

Then, we run simulations, and we see that it works, and that the representation is, at least, as good as GFT representation (but more stable).

Yes, as you correctly observe, the wavelet bank that we pick might come with a low-pass filter by themselves (like the Tight Hann wavelet when doing warping -when we do not do warping we don't have that-), in which case, we could, potentially, avoid having a low-pass summary operator, but we could use some other summary operator (it still needs to compute summaries), that could render better results. However, since we are comparing with other wavelets that do not have this low-pass operator, we decided to make the summary operator U, also a low-pass, and even though this might be redundant in the case of Tight Hann wavelets with warping, we see that it still works.

I hope this helps in clarifying. But feel free to keep asking if it was not clear or if you have any other questions!
